sysfs/class_power_supply.go

```diff
@@ -18,161 +18,275 @@ package sysfs
+	UsbType                  string // /sys/class/power_supply/<Name>/usb_type
+	VoltageAvg               *int64 // /sys/class/power_supply/<Name>/voltage_avg
+	VoltageBoot              *int64 // /sys/class/power_supply/<Name>/voltage_boot
+	VoltageMax               *int64 // /sys/class/power_supply/<Name>/voltage_max
+	VoltageMaxDesign         *int64 // /sys/class/power_supply/<Name>/voltage_max_design
+	VoltageMin               *int64 // /sys/class/power_supply/<Name>/voltage_min
+	VoltageMinDesign         *int64 // /sys/class/power_supply/<Name>/voltage_min_design
+	VoltageNow               *int64 // /sys/class/power_supply/<Name>/voltage_now
+	VoltageOCV               *int64 // /sys/class/power_supply/<Name>/voltage_ocv
 }
 
-// PowerSupplyClass is a collection of every power supply in /sys/class/power_supply/.
+// PowerSupplyClass is a collection of every power supply in
+// /sys/class/power_supply.
+//
 // The map keys are the names of the power supplies.
 type PowerSupplyClass map[string]PowerSupply
 
-// PowerSupplyClass returns info for all power supplies read from /sys/class/power_supply/.
+// PowerSupplyClass returns info for all power supplies read from
+// /sys/class/power_supply.
 func (fs FS) PowerSupplyClass() (PowerSupplyClass, error) {
 	path := fs.sys.Path("class/power_supply")
 
-	powerSupplyDirs, err := ioutil.ReadDir(path)
+	dirs, err := ioutil.ReadDir(path)
 	if err != nil {
-		return PowerSupplyClass{}, fmt.Errorf("cannot access %s dir %s", path, err)
+		return nil, fmt.Errorf("failed to list power supplies at %q: %v", path, err)
 	}
 
-	powerSupplyClass := PowerSupplyClass{}
-	for _, powerSupplyDir := range powerSupplyDirs {
-		powerSupply, err := powerSupplyClass.parsePowerSupply(path + "/" + powerSupplyDir.Name())
+	psc := make(PowerSupplyClass, len(dirs))
+	for _, d := range dirs {
+		ps, err := parsePowerSupply(filepath.Join(path, d.Name()))
 		if err != nil {
 			return nil, err
 		}
-		powerSupply.Name = powerSupplyDir.Name()
-		powerSupplyClass[powerSupplyDir.Name()] = *powerSupply
+
+		ps.Name = d.Name()
+		psc[d.Name()] = *ps
 	}
-	return powerSupplyClass, nil
+
+	return psc, nil
 }
 
-func (psc PowerSupplyClass) parsePowerSupply(powerSupplyPath string) (*PowerSupply, error) {
-	powerSupply := PowerSupply{}
-	powerSupplyElem := reflect.ValueOf(&powerSupply).Elem()
-	powerSupplyType := reflect.TypeOf(powerSupply)
+func parsePowerSupply(path string) (*PowerSupply, error) {
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
 
-	//start from 1 - skip the Name field
-	for i := 1; i < powerSupplyElem.NumField(); i++ {
-		fieldType := powerSupplyType.Field(i)
-		fieldValue := powerSupplyElem.Field(i)
+	var ps PowerSupply
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
 
-		if fieldType.Tag.Get("fileName") == "" {
-			panic(fmt.Errorf("field %s does not have a filename tag", fieldType.Name))
+		name := filepath.Join(path, f.Name())
+		value, err := util.SysReadFile(name)
```

Comment:
Yeah I think this could make sense and I was thinking about it yesterday, but I wanted to see what sort of usage patterns emerged after using this in a couple of places.

I think for now, I'd vote to leave this as-is and revisit in a future PR. I'd like to apply these same patterns for the network class stuff, and then we can refactor them all at once.

Reply:
Why do we open all files in the directory here and then later only process some based on the filename? Or am I missing something? Looks like this leads to https://github.com/prometheus/node_exporter/issues/2510